### PR TITLE
Pulse python bindings fork

### PR DESCRIPTION
**Pull request by mbhealy**
This PR adds support for python bindings of the qe-compiler's custom dialects (QUIR, Pulse, QCS, and OQ3). It includes support for a number of operations, but doesn't yet claim to provide full support for all of the operations of all dialects, rather, it is a proof of concept. A new pytest has been added to exercise the bindings as well.

This work was started by @bcdonovan, continued by Joshua Lou, and completed by me, and has been a long term goal of the project. 🎉 

Supersedes #141 because we have moved to the fork model of PRs.

### CODE REVIEW

File: lib/CAPI/Dialect/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# add_library(MLIRCAPIPulse Pulse.cpp)
```
Comment (bcdonovan) [inline lib/CAPI/Dialect/CMakeLists.txt:1]: Remove commented out lines?

**Comment by bcdonovan** [comment]
Rather than go through and mark them all a number of the file headers will need to be corrected for file name / description and should get the standard copyright date, Qiskit statement and license.

**Comment by mbhealy** [comment]
> Rather than go through and mark them all a number of the file headers will need to be corrected for file name / description and should get the standard copyright date, Qiskit statement and license.

👍 Will do.

File: python_lib/qss_compiler/examples/parse.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# Using this notebook requires manual install of MLIR python bindings:
```
Comment (bcdonovan) [inline python_lib/qss_compiler/examples/parse.py:4]: Are these comments about how to use the notebook still relevant?

Comment (bcdonovan) [reply python_lib/qss_compiler/examples/parse.py:4]: Same question for the notes in the example notebook.

File: python_lib/qss_compiler/examples/pulse.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        with InsertionPoint(module.body):
+            function = func.FuncOp("main", ([], [i32]))
+            function.add_entry_block()
+
+        with InsertionPoint(module.body):
+
+            seq1 = pulse.SequenceOp("seq_1", [wf, wf, mf, mf], [i1])
+            seq1.add_entry_block()
+
+        #     with InsertionPoint(seq1.entry_block):
```
Comment (bcdonovan) [inline python_lib/qss_compiler/examples/pulse.py:55]: Should these be un-commented or removed?

File: python_lib/setup_mlir.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
+alias sed=/opt/homebrew/opt/gsed/bin/gsed
```
Comment (bcdonovan) [inline python_lib/setup_mlir.sh:1]: This script was a testing setup script during initial development. If it is no longer used should it be removed?

**Review by bcdonovan — COMMENTED** [review]
A few small comments about some left over test / setup scripts and comments. 

Should we add a section to the readme about building / using the python bindings?

File: lib/API/api.cpp [hunk 5]
```diff
@@ -427,8 +427,6 @@ bool emitDiagnosticsAndCheckForErrors(
       case Severity::Warning:
       case Severity::Info:
         break;
-      default:
```
[the remark this answers is not in the document]
Comment (mbhealy) [reply lib/API/api.cpp]: Done in https://github.com/openqasm/qe-compiler/pull/264/commits/c2c3da735863574b2576e22b64f5848c0c8f2667

File: lib/CAPI/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    # add to include dirs
+    # set(_DIRS
+    #     ${MLIR_INCLUDE_DIRS}
+    #     ${PROJEC}
+    # )
+endfunction()
+
+add_subdirectory(Dialect)
+
+# Build the CAPI dylib.
```
[the remark this answers is not in the document]
Comment (mbhealy) [reply lib/CAPI/CMakeLists.txt]: Done in https://github.com/openqasm/qe-compiler/pull/264/commits/02bafaedd12067bd0eaffd7566ee37306f01e303

File: python_lib/qss_compiler/examples/parse.py (hunk 2 above)
Comment (mbhealy) [reply python_lib/qss_compiler/examples/parse.py]: Removed in https://github.com/openqasm/qe-compiler/pull/264/commits/02bafaedd12067bd0eaffd7566ee37306f01e303

File: python_lib/qss_compiler/examples/pulse.py (hunk 3 above)
Comment (mbhealy) [reply python_lib/qss_compiler/examples/pulse.py:62]: Removed in https://github.com/openqasm/qe-compiler/pull/264/commits/02bafaedd12067bd0eaffd7566ee37306f01e303

File: python_lib/setup_mlir.sh (hunk 4 above)
Comment (mbhealy) [reply python_lib/setup_mlir.sh]: Removed in https://github.com/openqasm/qe-compiler/pull/264/commits/8f588c934e1bd4b5af463a2a568b384d6d764690

**Comment by mbhealy** [comment]
Superseded by https://github.com/openqasm/qe-compiler/pull/276

**Pull request closed without merging**